ext/liquid_c/context.c

```diff
@@ -0,0 +1,129 @@
+
+    if (rb_respond_to(expression, id_evaluate))
+        return rb_funcall(expression, id_evaluate, 1, self);
+
+    return expression;
+}
+
+void context_maybe_raise_undefined_variable(VALUE self, VALUE key)
+{
+    if (rb_ivar_get(self, id_ivar_strict_variables)) {
+        Check_Type(key, T_STRING);
+        rb_enc_raise(utf8_encoding, cLiquidUndefinedVariable, "undefined variable %s", RSTRING_PTR(key));
+    }
+}
+
+VALUE context_find_variable(VALUE self, VALUE key, VALUE raise_on_not_found)
+{
+    VALUE scope = Qnil, variable = Qnil;
+
+    VALUE scopes = rb_ivar_get(self, id_ivar_scopes);
+    Check_Type(scopes, T_ARRAY);
+
+    for (long i = 0; i < RARRAY_LEN(scopes); i++) {
+        VALUE this_scope = RARRAY_AREF(scopes, i);
+        if (RB_LIKELY(TYPE(this_scope) == T_HASH)) {
+            // Does not invoke any default value proc, this is equivalent in
+            // cost and semantics to #key? but loads the value as well
+            variable = rb_hash_lookup2(this_scope, key, Qundef);
+            if (variable != Qundef) {
+                scope = this_scope;
+                goto variable_found;
+            }
+        } else if (RTEST(rb_funcall(this_scope, id_has_key, 1, key))) {
+            // Slow path: It is valid to pass a non-hash value to Liquid as a
+            // scope if it supports #key? and #[]
+            variable = rb_funcall(this_scope, id_aref, 1, key);
+            goto variable_found;
+        }
+    }
+
+    VALUE environments = rb_ivar_get(self, id_ivar_environments);
+    Check_Type(environments, T_ARRAY);
+    VALUE strict_variables = rb_ivar_get(self, id_ivar_strict_variables);
+
+    for (long i = 0; i < RARRAY_LEN(environments); i++) {
+        VALUE this_environ = RARRAY_AREF(environments, i);
+        if (RB_LIKELY(TYPE(this_environ) == T_HASH)) {
+            // Does not invoke any default value proc, this is equivalent in
+            // cost and semantics to #key? but loads the value as well
+            variable = rb_hash_lookup2(this_environ, key, Qundef);
+            if (variable != Qundef) {
+                scope = this_environ;
+                goto variable_found;
+            }
+
+            if (!(RTEST(raise_on_not_found) && RTEST(strict_variables))) {
+                // If we aren't running strictly, we need to invoke the default
+                // value proc, rb_hash_aref does this
+                variable = rb_hash_aref(this_environ, key);
+                if (variable != Qnil) {
+                    scope = this_environ;
+                    goto variable_found;
+                }
+            }
+        } else if (RTEST(rb_funcall(this_environ, id_has_key, 1, key))) {
+            // Slow path: It is valid to pass a non-hash value to Liquid as an
+            // environment if it supports #key? and #[]
+            variable = rb_funcall(this_environ, id_aref, 1, key);
+            goto variable_found;
+        }
+    }
+
+    if (RTEST(raise_on_not_found)) {
```

Comment:
Idea for later maybe: We could consider turning some of those Ruby runtime configuration flags (like strict mode, raise_on_not_found, etc.) into compile time flags (and `#ifdef` this code away). Might make things a tiny bit faster. Dunno if it's worth it. Probably not. Just a thought.

Reply:
Yeah I think it won't be worth it, since the typical path should be the non-error path.